pandas/core/apply.py

```diff
@@ -410,17 +410,32 @@ def agg_dict_like(self) -> DataFrame | Series:
             context_manager = com.temp_setattr(obj, "as_index", True)
         else:
             context_manager = nullcontext()
+
+        if isinstance(selected_obj, ABCDataFrame):
+            is_non_unique_col = selected_obj.columns.duplicated().tolist()
+        else:
+            is_non_unique_col = [False]
+
         with context_manager:
             if selected_obj.ndim == 1:
                 # key only used for output
-                colg = obj._gotitem(selection, ndim=1)
-                results = {key: colg.agg(how) for key, how in arg.items()}
+                key_res = obj._gotitem(selection, ndim=1)
+                results = {key: key_res.agg(how) for key, how in arg.items()}
+            elif any(is_non_unique_col):
+                # GH#51099
+                # results is a dict of lists
+                results = {}
+                for key, how in arg.items():
+                    key_res = []
+                    for col_idx in selected_obj.columns.get_indexer_for([key]):
```

Comment:
get_indexer_for is expensive, I think we should call it once. Something like

```
indices = selected_obj.columns.get_indexer_for(keys)
labels = selected_obj.columns.take(indices)
label_to_indices = defaultdict(list)
for index, label in zip(indices, labels):
    label_to_indices[label].append(index)
```

Then you can iterate through label_to_indices.